### PR TITLE
Add assert_required_keys to complement assert_valid_keys

### PR DESCRIPTION
**Pull request by corsonknowles**
### Summary
Adds a public method to `Hash` that can be used to ensure all required parameters are present before doing further work (e.g. during validations or at the beginning of an asynchronous job)

While `assert_valid_keys` ensures that no unexpected keys are present, this provides the complementary behavior, to ensure that no required keys are missing. 

Provides a helpful list of missing keys when the error is triggered. 

#### History
Attempts to implement in Rails
https://github.com/rails/rails/pull/10718/files
Attempts to implement outside of Rails:
https://github.com/laserlemon/figaro/commit/9f54872dfc1a972b4a971211706272f0f38495f4

#### Performance
TODO: add comparisons in approaches based on size of args * size of Hash object being operated on

### Other Information

TODO: 
Add to Changelog, like so: https://github.com/rails/rails/commit/62a20c2a642a2ae6ac7c265af1fe8fdf33114bf4#diff-69e4fa602a1f47160db003903d0de0671658b11a05dfe2883afe2988d429a032

### CODE REVIEW

**Comment by rafaelfranca** [comment]
Can you explain why you want to introduce this new method? I understand what it does, but I don't understand in which cases it would be useful.

**Comment by corsonknowles** [comment]
> Can you explain why you want to introduce this new method? I understand what it does, but I don't understand in which cases it would be useful.

Yes, great question! I maintain a suite of Rails applications that all use this method. It can be very useful, just for example with ActiveJob or Sidekiq, to have required keys that must be present for the job to run.

I thought it would be kind to push this upstream as a general utility that complements the functionality of `assert_valid_keys`

I'm wondering if the usefulness of this could/should be demonstrated by putting it to work in the Rails codebase -- perhaps here:
https://github.com/rails/rails/blob/master/actionpack/lib/action_dispatch/http/upload.rb#L29
or in Router where a similar pattern is used.

**Comment by rafaelfranca** [comment]
Would not required keyword arguments be enough for most cases? In the case on https://github.com/rails/rails/blob/master/actionpack/lib/action_dispatch/http/upload.rb#L29 I actually prefer the current implementation since it will not do another operation in the hash.

**Comment by corsonknowles** [comment]
Yep, I agree --- looking at that action dispatch method as an example, I was trying to find a way to use this with equal or greater performance, but it's just fine as is.

Required keyword arguments is similar and covers a lot of the need, but doesn't cover the case of operations on a hash of parameters unless you refactor to create an explicit method with matching keyword arguments for each place you need to do this.

This is my first time trying to contribute up to Rails, so requests and advice on how to show this is useful are very welcome. 

I can tell you that we use this method in our base library (I decided to open this PR after doing some benchmarking to demonstrate that the `.key?` implementation is generally preferable to something like `missing_keys = required_keys - self.keys` and promoting it from our Rails base library to our Ruby base library to de-duplicate multiple declarations of the same method). 

We're a pretty large Ruby shop, larger than Github but currently smaller than AirBnB, and we think this is a core piece of utility to provide for hashes. Let me know what the burden of proof is to merit adding this to Rails 6.2, and I will see if I can help meet it. 

I tried to do a bit of a lit review to see who else is already doing something like this:

The author of Flipper casually implementing a similar method with the same name:
https://github.com/jnunemaker/crdts/blob/1da567fe488840ad91ab6372a7996e04d6410f34/lib/crdts/converters/hash.rb#L26

The Ruby Github API library:
https://github.com/piotrmurach/github/blob/02ec883ee7038385559854c1081d235cb2dd1245/lib/github_api/validations/required.rb#L14

Official Ruby Bitbucket API:
https://github.com/bitbucket-rest-api/bitbucket/blob/2fc8fa575ca3636ee8603a4b0efce95bbd79fab0/lib/bitbucket_rest_api/validations/required.rb

**Comment by rafaelfranca** [comment]
We have a policy of not adding new methods to Ruby core classes unless it would benefit the framework implementation or the large majority of applications. For most cases I believe the required keyword arguments are enough. This is actually the case for `assert_valid_keys` as well since we can change most calls to this method to keyword arguments.

To introduce this to 6.2 we need to show that this would be useful for the framework, and only two places don't are not enough. If you want to investigate where we could use this method in the framework it would help.

File: activesupport/lib/active_support/core_ext/hash/keys.rb [hunk 1]
```diff
@@ -54,6 +54,20 @@ def assert_valid_keys(*valid_keys)
...
+  # +ArgumentError+ for any missing keys.
+  #
+  # Note that keys are treated differently than HashWithIndifferentAccess,
+  # meaning that string and symbol keys will not match.
+  #
+  #   { name: 'Rob', age: '28' }.assert_required_keys('name', 'age') # => raises "ArgumentError: Missing required option(s): 'name', 'age'"
+  #   { name: 'Rob', years: '28' }.assert_required_keys(:name, :age) # => raises "ArgumentError: Missing required option(s): :age"
+  #   { name: 'Rob', age: '28' }.assert_required_keys(:name, :age)   # => passes, raises nothing
+  def assert_required_keys(*required_keys)
+    missing_keys = required_keys.reject { |key| key?(key) }
```
Comment (corsonknowles) [inline activesupport/lib/active_support/core_ext/hash/keys.rb:67]:
```suggestion
    missing_keys = required_keys.flatten.reject { |key| key?(key) }
```

File: activesupport/lib/active_support/core_ext/hash/keys.rb [hunk 2]
```diff
@@ -54,6 +54,20 @@ def assert_valid_keys(*valid_keys)
...
+  #
+  # Note that keys are treated differently than HashWithIndifferentAccess,
+  # meaning that string and symbol keys will not match.
+  #
+  #   { name: 'Rob', age: '28' }.assert_required_keys('name', 'age') # => raises "ArgumentError: Missing required option(s): 'name', 'age'"
+  #   { name: 'Rob', years: '28' }.assert_required_keys(:name, :age) # => raises "ArgumentError: Missing required option(s): :age"
+  #   { name: 'Rob', age: '28' }.assert_required_keys(:name, :age)   # => passes, raises nothing
+  def assert_required_keys(*required_keys)
+    missing_keys = required_keys.flatten.reject { |key| key?(key) }
+    raise ArgumentError.new("Missing required option(s): #{missing_keys.join(", ")}") unless missing_keys.empty?
```
Comment (corsonknowles) [inline activesupport/lib/active_support/core_ext/hash/keys.rb:68]:
```suggestion
    raise ArgumentError.new("Missing required option(s): #{missing_keys.map(&:inspect).join(", ")}") unless missing_keys.empty?
```

**Comment by corsonknowles** [comment]
I'm going to close this for now. Happy to revisit when we can identify some internal callers of this method. Thanks @rafaelfranca for the excellent comments

**Pull request closed without merging**